### PR TITLE
310 idea portal refactoring

### LINKED ISSUE #310: Idea Portal refactoring
should display less before hackathon selected
Participate needs feedback
Participate button uses participant id in request instead of user id
Participant button never goes away
should be gone when user has a participant for a hackathon
changing hackathon in dropdown before loading of data complete doesn't cancel operation or start a new one to get data
Idea Details
description never expands
participant heads should disappear when you expand the participant details

### PR DESCRIPTION
**Pull request by LetsPoke**
merge #327 first !

- should display less before hackathon selected ✅
- Participate needs feedback ✅
- Participate button uses participant id in request instead of user id ✅
- Participant button never goes away ✅
- should be gone when user has a participant for a hackathon ✅ gets replaced with a "left hackathon" button
- Idea Details description never expands ✅
- participant heads should disappear when you expand the participant details ✅
- participate button should update without reloading page ✅

closes #310

### CODE REVIEW

File: hms-app/src/pages/IdeaPortal.tsx [hunk 1]
```diff
@@ -38,14 +43,77 @@ function IdeaPortal() {
...
+      title: 'Join Hackathon',
+      message: 'this can take a second',
+      autoClose: false,
+      disallowClose: true,
+    })
     createParticipant(participantInfo.userId, participantInfo.hackathonId).then(
       (r) => {
-        console.log(r)
+        setTimeout(() => {
+          console.log('r added', r)
```
Comment (verzano) [inline hms-app/src/pages/IdeaPortal.tsx:69]:
```suggestion
          console.log('Participant Added', r)
```

File: hms-app/src/pages/Voting.tsx [hunk 2]
```diff
@@ -188,7 +187,9 @@ export default function Voting() {
                             {column.items.map((item, index) => {
                               return (
                                 <Draggable
+                                  // @ts-ignore
                                   key={item.id}
+                                  // @ts-ignore
                                   draggableId={item.id.toString()}
```
Comment (verzano) [inline hms-app/src/pages/Voting.tsx:193]:
```suggestion
                                  draggableId={item.id}
```

File: hms-app/src/pages/IdeaPortal.tsx [hunk 3]
```diff
@@ -38,14 +43,77 @@ function IdeaPortal() {
...
+      autoClose: false,
+      disallowClose: true,
+    })
+    // @ts-ignore
+    deleteParticipant(filterParticipants[0].id).then((r) => {
+      setTimeout(() => {
+        console.log('r deleted', r)
+        setButtonisDisabled(false)
+        setParticipantCheck(false)
+        updateNotification({
```
Comment (verzano) [inline hms-app/src/pages/IdeaPortal.tsx:105]:
```suggestion
      console.log('r deleted', r)
      setButtonisDisabled(false)
      setParticipantCheck(false)
      setTimeout(() => {
        updateNotification({
```

**Review by verzano — APPROVED** [review]

File: hms-app/src/pages/Voting.tsx [hunk 4]
```diff
@@ -188,7 +187,9 @@ export default function Voting() {
                             {column.items.map((item, index) => {
                               return (
                                 <Draggable
+                                  // @ts-ignore
```
Comment (KBernaszuk) [inline hms-app/src/pages/Voting.tsx:190]: I couldn't go through it yesterday, but the reason that you need to add those ignores, is unclear structure of your variables which can lead to - undefined, errors and other bad things. The best option would be to never add such annotations and use strong typing. 😄 

You were missing, at the def. level:
```
const columnsFromBackend : { [key: string]: {name: string, items: {id: string}[]}; }= {
  ['1']: {
    name: 'All ideas',
    items: [],
  },
  ['2']: {
    name: 'Your Votes',
    items: [],
  },
}
```

Comment (KBernaszuk) [reply hms-app/src/pages/Voting.tsx:190]: or simpler like:
```
type Item = {id: string}

const columnsFromBackend = {
  ['1']: {
    name: 'All ideas',
    items: [] as Item[],
  },
  ['2']: {
    name: 'Your Votes',
    items: [],
  },
}
```

**Pull request merged**